### PR TITLE
[FLINK-25755][runtime] AdaptiveScheduler: Introduce shared interface…

### PR DESCRIPTION
**Pull request by dmvk**
https://issues.apache.org/jira/browse/FLINK-25755

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateTransitions.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An interface covering all possible {@link State} transitions. The main purpose is to align the
+ * transition methods between different contexts.
+ */
+public interface StateTransitions {
+
+    interface ToCancelling extends StateTransitions {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateTransitions.java:36]: These interface are missing javadocs.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java [hunk 2]
```diff
@@ -271,18 +264,17 @@ public AdaptiveScheduler(
 
         this.executionGraphFactory = executionGraphFactory;
 
+        final JobStatusStore jobStatusStore = new JobStatusStore(initializationTimestamp);
         final Collection<JobStatusListener> tmpJobStatusListeners = new ArrayList<>();
         tmpJobStatusListeners.add(Preconditions.checkNotNull(jobStatusListener));
         tmpJobStatusListeners.add(jobStatusStore);
-
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:277]: what are these empty-line changes about?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java [hunk 3]
```diff
@@ -346,14 +346,6 @@ abstract boolean updateTaskExecutionState(
...
-        Executor getMainThreadExecutor();
-
-        /**
-         * Transitions into the {@link Finished} state.
-         *
-         * @param archivedExecutionGraph archivedExecutionGraph which is passed to the {@link
-         *     Finished} state
-         */
-        void goToFinished(ArchivedExecutionGraph archivedExecutionGraph);
+        ComponentMainThreadExecutor getMainThreadExecutor();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:349]: Why is this now using a different return type?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java [hunk 4]
```diff
@@ -227,6 +237,16 @@ public void goToWaitingForResources() {
             hadStateTransitionHappened = true;
         }
 
+        @Override
+        public Executor getIOExecutor() {
+            return MoreExecutors.directExecutor();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraphTest.java:242]:
```suggestion
            return Executors.directExecutor();
```
from org.apache.flink.util.concurrent.Executors

**Review by zentol — CHANGES_REQUESTED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java (hunk 2 above)
Comment (dmvk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java]: removed

**Review by dmvk — COMMENTED** [review]
Thanks for the review @zentol , ready for another pass

**Pull request merged**